### DISCUSSION TITLE
Empty list of quantity sets in IFC-file (Ifcopenshell)

### DISCUSSION THREAD

**Discussion by elisemunche** [IfcOpenShell/IfcOpenShell#2914] [Q&A]
Hi! 

I am trying to write a code that extracts the quantity sets from an IFC-file for IfcBeam and IfcColumn. 
I have this code that works perfectly for most of the IFC files I have tried: 

import ifcopenshell 
from ifcopenshell import util 

ifc_path = r"C:\Users\elisemun\Documents\GitHub\vsc_test\test_ifc_revit.ifc"
ifc_file = ifcopenshell.open(ifc_path)

all_elements=ifc_file.by_type("IfcBuildingElement")

#Wanted element types 
LB_elements = [] 
for element_type in ['IfcBeam', 'IfcColumn']:
    LB_elements.extend(ifc_file.by_type(element_type))

#Qtos (Quantity set)
info_elements= [ifcopenshell.util.element.get_psets(element,qtos_only=True) for element in LB_elements]

However, I have some IFC files that give me the result info_elements=[ {},{},{}.....{}]. I assume this means that there is no value here with the method I am using. When looking in the IFC files that do not work with my code, I cant seem to find the entities (amongst others): IfcQuantityArea and IfcQuantityLength, like I can in the files that do work. 

So my question is, does anyone know how I can reach the quantities in such a file by using Ifcopenshell? 

Thank you!

**Comment by aothms** [comment]
Quantities (and property sets) are not guaranteed to populated in a model (or generated consistently for that matter). Luckily ifcopenshell provides the tooling to generate a lot of quantities from the geometry. https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/util/shape.py

**Reply by elisemunche** [reply]
I see! Thank you so much for the response, I am looking into it now :D 
Do you also by chance know what is meant by "geometry" as requirements in several of the definitions? @aothms

**Reply by aothms** [reply]
The result of geom.create_shape() / geom.iterator.get (with USE_PYTHON_OPENCASCADE to false, which is the default).

```python
>>> import ifcopenshell.geom
>>> import ifcopenshell.util.shape
>>> f = ifcopenshell.open('acad2010_walls.ifc')
>>> elem = ifcopenshell.geom.create_shape(ifcopenshell.geom.settings(), f.by_type('IfcElement')[0])
>>> ifcopenshell.util.shape.get_volume(elem.geometry)
0.9425983811263173
```

**Comment by elisemunche** [comment]
I see, thank you!
I have been playing around with the code for a while now but cant seem to understand how to extract the cross section areas and the length of the elements using the library. May I ask if you have any ideas?

**Reply by aothms** [reply]
It really depends on the kinds of elements. Typical walls will have their length around the local x axis, height along long z axis. For other elements other conventions exist. So you can get quite far with looking min max values of coordinate components in the mesh. 

It all breaks down of course when you start looking at curved elements or anything that deviates from the convention. Typical walls will have an axis representation (2d polyline or arc segment) that you could use to determine the length even in non-standard situations.

In other cases it makes sense to look at the extrusion (IfcExtrudedAreaSolid) in IFC or reverse engineer an explicit mesh into an extrusion. @Moult surely has some code for this.

**Reply by elisemunche** [reply]
Ah okei. Both me and my thesis are thankful for you taking the time! 
I am mainly looking at IfcBeams and IfcColumns. @Moult, do you have any tips on how I can do this for these elements?

**Reply by arunkvijay** [reply]
Hi @aothms How do I get the `Depth` from `IfcExtrudedAreaSolid` of IfcColumn element programmatically

**Reply by aothms** [reply]
All attributes are available on the entity instances, so look at Representation Representations Items Depth, or loop over `ifcopenshell.file.traverse(my_column)` until you find what you need

**Reply by Moult** [reply]
I've seen this come up in a few places on how to get extrusions specifically (they are one of the most common representation items) or cross sectional profiles so this might help https://github.com/IfcOpenShell/IfcOpenShell/commit/260eb31a695731dabcd729db2510d75e24905a28

**Reply by aothms** [reply]
That's great as it also handles boolean operands which I didn't discuss

**Reply by arunkvijay** [reply]
> I've seen this come up in a few places on how to get extrusions specifically (they are one of the most common representation items) or cross sectional profiles so this might help [260eb31](https://github.com/IfcOpenShell/IfcOpenShell/commit/260eb31a695731dabcd729db2510d75e24905a28)

Thanks @Moult, this is very helpful. There is a small spelling mistake in L301 (to be changed to extrusions). Should I submit a PR for this?

**Reply by aothms** [reply]
@arunkvijay yes please :) according to the interpreter that's not a harmless spelling mistake, but a NameError. Good catch :)
